### PR TITLE
gh-108220: Internal header files require Py_BUILD_CORE to be defined

### PR DESCRIPTION
**Pull request by vstinner**
* pycore_intrinsics.h does nothing if included twice (add #ifndef and #define).
* Update Tools/cases_generator/generate_cases.py to generate the Py_BUILD_CORE test.

* Issue: gh-108220

### CODE REVIEW

File: Tools/cases_generator/generate_cases.py [hunk 1]
```diff
@@ -403,6 +404,12 @@ def write_metadata(self, metadata_filename: str, pymetadata_filename: str) -> No
 
             self.write_provenance_header()
 
+            self.out.emit("\n" + textwrap.dedent("""
+                #ifndef Py_BUILD_CORE
+                #  error "this header requires Py_BUILD_CORE define"
+                #endif
+            """).strip())
```
Comment (gvanrossum) [inline Tools/cases_generator/generate_cases.py:411]: Why bother asking for a review if you're going to merge it while I'm getting coffee? What response time do you expect for reviews?

Comment (corona10) [reply Tools/cases_generator/generate_cases.py:411]: > Why bother asking for a review

Hmm, I requested the review to you.

Comment (gvanrossum) [reply Tools/cases_generator/generate_cases.py:411]: Yeah,  I was venting at @vstinner who ignored that and merged **eight minutes later**. :-(

Comment (vstinner) [reply Tools/cases_generator/generate_cases.py:411]: I didn't ask for a review on such "cleanup" change. I didn't see that @corona10 asked for a review.

**Pull request merged**